File: platform/script-debugger/protocol/schema-reader-generator/src/org/jetbrains/jsonProtocol/ProtocolMetaModel.java
package org.jetbrains.jsonProtocol;

import org.chromium.protocolReader.JsonField;
import org.chromium.protocolReader.JsonOptionalField;
import org.chromium.protocolReader.JsonType;

import java.util.List;

/**
 * Defines schema of WIP metamodel defined in http://svn.webkit.org/repository/webkit/trunk/Source/WebCore/inspector/Inspector.json
 */
public interface ProtocolMetaModel {
  @JsonType
  interface Root {
    @JsonOptionalField
    Version version();

    List<Domain> domains();
  }

  @JsonType
  interface Version {
    String major();
    String minor();
  }

  @JsonType
  interface Domain {
    String domain();

    @JsonOptionalField
    List<StandaloneType> types();

    List<Command> commands();

    @JsonOptionalField
    List<Event> events();

    @JsonOptionalField
    String description();

    @JsonOptionalField
    boolean hidden();
  }

  @JsonType
  interface Command {
    String name();

    @JsonOptionalField
    List<Parameter> parameters();

    @JsonOptionalField
    List<Parameter> returns();

    @JsonOptionalField
    String description();

    @JsonOptionalField
    boolean hidden();

    @JsonOptionalField
    boolean async();
  }

  @JsonType
  interface Parameter extends ItemDescriptor.Named {
    @Override
    String name();

    @Override
    @JsonOptionalField
    String shortName();

    @Override
    @JsonOptionalField
    String type();

    @Override
    @JsonOptionalField
    ArrayItemType items();

    @Override
    @JsonField(jsonLiteralName = "enum")
    @JsonOptionalField
    List<String> getEnum();

    @Override
    @JsonOptionalField
    @JsonField(jsonLiteralName = "$ref")
    String ref();

    @Override
    @JsonOptionalField
    boolean optional();

    @Override
    @JsonOptionalField
    String description();

    @JsonOptionalField
    boolean hidden();
  }

  @JsonType
  interface Event {
    String name();

    @JsonOptionalField
    List<Parameter> parameters();

    @JsonOptionalField
    String description();

    @JsonOptionalField
    boolean hidden();
  }

  @JsonType
  interface StandaloneType extends ItemDescriptor.Type {
    String id();

    @Override
    @JsonOptionalField
    String description();

    @Override
    String type();

    @JsonOptionalField
    boolean hidden();

    @Override
    @JsonOptionalField
    List<ObjectProperty> properties();

    @Override
    @JsonField(jsonLiteralName = "enum")
    @JsonOptionalField
    List<String> getEnum();

    @Override
    @JsonOptionalField
    ArrayItemType items();
  }


  @JsonType
  interface ArrayItemType extends ItemDescriptor.Type, ItemDescriptor.Referenceable {
    @Override
    @JsonOptionalField
    String description();

    @JsonOptionalField
    boolean optional();

    @Override
    @JsonOptionalField
    String type();

    @Override
    @JsonOptionalField
    ArrayItemType items();

    @Override
    @JsonField(jsonLiteralName = "$ref")
    @JsonOptionalField
    String ref();

    @Override
    @JsonField(jsonLiteralName = "enum")
    @JsonOptionalField
    List<String> getEnum();

    @Override
    @JsonOptionalField
    List<ObjectProperty> properties();
  }

  @JsonType
  interface ObjectProperty extends ItemDescriptor.Named {
    @Override
    String name();

    @Override
    @JsonOptionalField
    String shortName();

    @Override
    @JsonOptionalField
    String description();

    @Override
    @JsonOptionalField
    boolean optional();

    @Override
    @JsonOptionalField
    String type();

    @Override
    @JsonOptionalField
    ArrayItemType items();

    @Override
    @JsonField(jsonLiteralName = "$ref")
    @JsonOptionalField
    String ref();

    @Override
    @JsonField(jsonLiteralName = "enum")
    @JsonOptionalField
    List<String> getEnum();

    @JsonOptionalField
    boolean hidden();
  }

  String STRING_TYPE = "string";
  String INTEGER_TYPE = "integer";
  String NUMBER_TYPE = "number";
  String BOOLEAN_TYPE = "boolean";
  String OBJECT_TYPE = "object";
  String ARRAY_TYPE = "array";
  String UNKNOWN_TYPE = "unknown";
  String ANY_TYPE = "any";
}

File: platform/script-debugger/protocol/schema-reader-generator/src/ProtocolMetaModel.kt
// Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.jetbrains.jsonProtocol

import org.jetbrains.io.JsonReaderEx

const val STRING_TYPE: String = "string"
const val INTEGER_TYPE: String = "integer"
const val NUMBER_TYPE: String = "number"
const val BOOLEAN_TYPE: String = "boolean"
public const val OBJECT_TYPE: String = "object"
const val ARRAY_TYPE: String = "array"
const val UNKNOWN_TYPE: String = "unknown"
const val ANY_TYPE: String = "any"

interface ItemDescriptor {
  val description: String?

  val type: String?

  val enum: List<String>?

  val items: ProtocolMetaModel.ArrayItemType?

  interface Named : Referenceable {
    fun name(): String

    val shortName: String?

    val optional: Boolean
  }

  interface Referenceable : ItemDescriptor {
    @ProtocolName("\$ref")
    val ref: String?
  }

  interface Type : ItemDescriptor {
    val properties: List<ProtocolMetaModel.ObjectProperty>?
  }
}

interface ProtocolSchemaReader {
  @JsonParseMethod
  fun parseRoot(reader: JsonReaderEx): ProtocolMetaModel.Root
}

/**
 * Defines schema of WIP metamodel defined in http://svn.webkit.org/repository/webkit/trunk/Source/WebCore/inspector/Inspector.json
 */
interface ProtocolMetaModel {
  @JsonType
  interface Root {
    val version: Version?

    fun domains(): List<Domain>
  }

  interface Version {
    fun major(): String
    fun minor(): String
  }

  interface Domain {
    fun domain(): String

    val types: List<StandaloneType>?

    fun commands(): List<Command>

    val events: List<Event>?

    val description: String?

    val hidden: Boolean

    val experimental: Boolean
  }

  interface Command {
    fun name(): String

    val parameters: List<Parameter>?

    val returns: List<Parameter>?

    val description: String?

    val hidden: Boolean

    val async: Boolean
  }

  interface Parameter : ItemDescriptor.Named {
    val hidden: Boolean

    @JsonField(allowAnyPrimitiveValue = true)
    val default: String?
  }

  interface Event {
    fun name(): String

    val parameters: List<Parameter>?

    val description: String?

    val hidden: Boolean

    val optionalData: Boolean
  }

  @JsonType
  interface StandaloneType : ItemDescriptor.Type {
    fun id(): String

    val hidden: Boolean
  }

  interface ArrayItemType : ItemDescriptor.Type, ItemDescriptor.Referenceable {
    val optional: Boolean
  }

  interface ObjectProperty : ItemDescriptor.Named {
    override fun name(): String

    val hidden: Boolean
  }
}